File: apps/api/services/billing_customer_dispatch.py
from __future__ import annotations

import asyncio
import time
from contextlib import suppress
from dataclasses import dataclass

import httpx
from sqlalchemy import or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from core.config import Settings, get_settings
from core.logging import get_logger
from core.timestamps import now_unix_seconds
from db.models import BillingCustomerOutbox
from db.session import AsyncSessionLocal
from services.billing_customer_sync import customer_event_payload

logger = get_logger(__name__)

_LOCK_TIMEOUT_SECONDS = 5 * 60
_MAX_ERROR_LENGTH = 2_000


@dataclass(frozen=True)
class BillingCustomerDispatchSummary:
    claimed: int
    delivered: int
    failed: int
    configured: bool


async def claim_billing_customer_events(
    db: AsyncSession,
    *,
    now: int,
    limit: int,
) -> list[BillingCustomerOutbox]:
    retryable = or_(
        BillingCustomerOutbox.status.in_(("pending", "failed")),
        (
            (BillingCustomerOutbox.status == "processing")
            & (BillingCustomerOutbox.locked_at <= now - _LOCK_TIMEOUT_SECONDS)
        ),
    )
    rows = list(
        (
            await db.scalars(
                select(BillingCustomerOutbox)
                .where(retryable, BillingCustomerOutbox.available_at <= now)
                .order_by(BillingCustomerOutbox.created_at, BillingCustomerOutbox.id)
                .with_for_update(skip_locked=True)
                .limit(limit)
            )
        ).all()
    )
    for row in rows:
        row.status = "processing"
        row.attempt_count += 1
        row.locked_at = now
        row.last_error = None
        row.updated_at = now
    await db.flush()
    return rows


async def dispatch_billing_customer_sync_once(
    settings: Settings | None = None,
) -> BillingCustomerDispatchSummary:
    active_settings = settings or get_settings()
    billing_url = active_settings.billing_url.rstrip("/")
    if not billing_url or not active_settings.billing_internal_key:
        return BillingCustomerDispatchSummary(claimed=0, delivered=0, failed=0, configured=False)

    async with AsyncSessionLocal() as session:
        rows = await claim_billing_customer_events(
            session,
            now=now_unix_seconds(),
            limit=active_settings.finance_provisioning_batch_size,
        )
        snapshots = [(row.id, row.attempt_count, customer_event_payload(row)) for row in rows]
        await session.commit()

    if not snapshots:
        return BillingCustomerDispatchSummary(claimed=0, delivered=0, failed=0, configured=True)

    delivered = 0
    failed = 0
    endpoint = f"{billing_url}/api/v1/admin/customers/ensure"
    async with httpx.AsyncClient(timeout=httpx.Timeout(15.0)) as client:
        for event_id, attempt_count, payload in snapshots:
            try:
                response = await client.post(
                    endpoint,
                    headers={
                        "x-internal-key": active_settings.billing_internal_key,
                        "content-type": "application/json",
                        "x-request-id": event_id,
                    },
                    json=payload,
                )
                response.raise_for_status()
            except Exception as exc:
                failed += 1
                error = _delivery_error(exc)
                await _mark_failed(event_id, attempt_count, error)
                logger.warning(
                    "billing_customer_sync.delivery_failed",
                    event_id=event_id,
                    attempt_count=attempt_count,
                    error=error,
                )
            else:
                delivered += 1
                await _mark_delivered(event_id)
                logger.info(
                    "billing_customer_sync.delivered",
                    event_id=event_id,
                    attempt_count=attempt_count,
                )

    return BillingCustomerDispatchSummary(
        claimed=len(snapshots),
        delivered=delivered,
        failed=failed,
        configured=True,
    )


async def trigger_billing_run_once(settings: Settings) -> bool:
    billing_url = settings.billing_url.rstrip("/")
    if not billing_url or not settings.billing_internal_key:
        return False
    try:
        async with httpx.AsyncClient(timeout=httpx.Timeout(30.0)) as client:
            response = await client.post(
                f"{billing_url}/api/v1/admin/billing/run",
                headers={"x-internal-key": settings.billing_internal_key},
            )
            response.raise_for_status()
    except Exception as exc:
        logger.warning("billing_run.trigger_failed", error=_delivery_error(exc))
        return False
    return True


async def run_billing_sync_worker(
    stop: asyncio.Event,
    settings: Settings | None = None,
) -> None:
    active_settings = settings or get_settings()
    last_billing_run = time.monotonic()
    while not stop.is_set():
        try:
            await dispatch_billing_customer_sync_once(active_settings)
            now_monotonic = time.monotonic()
            if (
                active_settings.billing_run_interval_seconds > 0
                and now_monotonic - last_billing_run >= active_settings.billing_run_interval_seconds
            ):
                last_billing_run = now_monotonic
                await trigger_billing_run_once(active_settings)
        except Exception:
            logger.error("billing_customer_sync.worker_failed", exc_info=True)

        with suppress(TimeoutError):
            await asyncio.wait_for(
                stop.wait(),
                timeout=active_settings.finance_provisioning_poll_seconds,
            )


async def _mark_delivered(event_id: str) -> None:
    async with AsyncSessionLocal() as session:
        event = await session.get(BillingCustomerOutbox, event_id, with_for_update=True)
        if event is None or event.status != "processing":
            return
        now = now_unix_seconds()
        event.status = "delivered"
        event.delivered_at = now
        event.locked_at = None
        event.last_error = None
        event.updated_at = now
        await session.commit()


async def _mark_failed(event_id: str, attempt_count: int, message: str) -> None:
    async with AsyncSessionLocal() as session:
        event = await session.get(BillingCustomerOutbox, event_id, with_for_update=True)
        if event is None or event.status != "processing":
            return
        now = now_unix_seconds()
        event.status = "failed"
        event.available_at = now + min(3_600, 5 * (2 ** min(attempt_count, 10)))
        event.locked_at = None
        event.last_error = message[:_MAX_ERROR_LENGTH]
        event.updated_at = now
        await session.commit()


def _delivery_error(error: Exception) -> str:
    if isinstance(error, httpx.HTTPStatusError):
        body = error.response.text[:500].strip()
        return f"Billing returned HTTP {error.response.status_code}: {body}"
    return f"{type(error).__name__}: {error}"

File: apps/api/src/workers/billing-customer-dispatch.ts
import { getSettings } from '@/config'
import { getLogger } from '@/platform/logger'
import { nowUnixSeconds } from '@/platform/timestamps'
import { customerEventPayload } from '@/services/billing-customer-sync'

import {
  claimBillingCustomerEvents,
  markBillingCustomerDelivered,
  markBillingCustomerFailed,
} from './billing-customer-dispatch.repository'

const logger = getLogger('billing-customer-dispatch')

export type BillingCustomerDispatchSummary = {
  claimed: number
  delivered: number
  failed: number
  configured: boolean
}

function deliveryError(error: unknown): string {
  if (
    error instanceof Error &&
    error.message.startsWith('Billing returned HTTP')
  ) {
    return error.message.slice(0, 2000)
  }
  if (error instanceof Error) {
    return `${error.name}: ${error.message}`.slice(0, 2000)
  }
  return `${String(error)}`.slice(0, 2000)
}

async function postWithTimeout(
  url: string,
  headers: Record<string, string>,
  body: unknown,
  timeoutMs: number
): Promise<Response> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeoutMs)
  try {
    return await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
      signal: controller.signal,
    })
  } finally {
    clearTimeout(timer)
  }
}

export async function dispatchBillingCustomerSyncOnce(): Promise<BillingCustomerDispatchSummary> {
  const settings = getSettings()
  const billingUrl = settings.billing.url.trim().replace(/\/+$/, '')
  const internalKey = settings.billing.internalKey.trim()

  if (!billingUrl || !internalKey) {
    return { claimed: 0, delivered: 0, failed: 0, configured: false }
  }

  const now = nowUnixSeconds()
  const limit = settings.billing.financeProvisioningBatchSize

  const claimedRows = await claimBillingCustomerEvents(now, limit)

  if (claimedRows.length === 0) {
    return { claimed: 0, delivered: 0, failed: 0, configured: true }
  }

  // Snapshot before releasing the transaction lock — mirrors Python's
  // `snapshots = [(row.id, row.attempt_count, payload) for row in rows]`.
  const snapshots = claimedRows.map((row) => ({
    id: row.id,
    attemptCount: row.attemptCount,
    payload: customerEventPayload(row as never),
  }))

  let delivered = 0
  let failed = 0
  const endpoint = `${billingUrl}/api/v1/admin/customers/ensure`

  for (const item of snapshots) {
    try {
      const response = await postWithTimeout(
        endpoint,
        {
          'x-internal-key': internalKey,
          'content-type': 'application/json',
          'x-request-id': item.id,
        },
        item.payload,
        15_000
      )
      if (!response.ok) {
        const text = await response.text()
        const snippet = text.slice(0, 500).trim()
        throw new Error(`Billing returned HTTP ${response.status}: ${snippet}`)
      }
      delivered += 1
      await markBillingCustomerDelivered(item.id, nowUnixSeconds())
      logger.info(
        { event_id: item.id, attempt_count: item.attemptCount },
        'billing_customer_sync.delivered'
      )
    } catch (error) {
      failed += 1
      const message = deliveryError(error)
      await markBillingCustomerFailed(
        item.id,
        item.attemptCount,
        message,
        nowUnixSeconds()
      )
      logger.warn(
        { event_id: item.id, attempt_count: item.attemptCount, error: message },
        'billing_customer_sync.delivery_failed'
      )
    }
  }

  return { claimed: snapshots.length, delivered, failed, configured: true }
}

export async function triggerBillingRunOnce(): Promise<boolean> {
  const settings = getSettings()
  const billingUrl = settings.billing.url.trim().replace(/\/+$/, '')
  const internalKey = settings.billing.internalKey.trim()
  if (!billingUrl || !internalKey) return false

  try {
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), 30_000)
    let response: Response
    try {
      response = await fetch(`${billingUrl}/api/v1/admin/billing/run`, {
        method: 'POST',
        headers: { 'x-internal-key': internalKey },
        signal: controller.signal,
      })
    } finally {
      clearTimeout(timer)
    }
    if (!response.ok) {
      const text = await response.text()
      const snippet = text.slice(0, 500).trim()
      throw new Error(`Billing returned HTTP ${response.status}: ${snippet}`)
    }
    return true
  } catch (error) {
    logger.warn({ error: deliveryError(error) }, 'billing_run.trigger_failed')
    return false
  }
}

function sleepWithAbort(ms: number, signal?: AbortSignal): Promise<void> {
  if (!signal) return new Promise((resolve) => setTimeout(resolve, ms))
  if (signal.aborted) return Promise.resolve()
  return new Promise<void>((resolve) => {
    const timeout = setTimeout(() => {
      signal.removeEventListener('abort', onAbort)
      resolve()
    }, ms)
    const onAbort = (): void => {
      clearTimeout(timeout)
      signal.removeEventListener('abort', onAbort)
      resolve()
    }
    signal.addEventListener('abort', onAbort, { once: true })
  })
}

export async function runBillingSyncWorker(options?: {
  signal?: AbortSignal
}): Promise<void> {
  const settings = getSettings()
  let lastBillingRun = performance.now() / 1000

  while (!options?.signal?.aborted) {
    try {
      await dispatchBillingCustomerSyncOnce()
      const nowMono = performance.now() / 1000
      if (
        settings.billing.runIntervalSeconds > 0 &&
        nowMono - lastBillingRun >= settings.billing.runIntervalSeconds
      ) {
        lastBillingRun = nowMono
        await triggerBillingRunOnce()
      }
    } catch (error) {
      logger.error({ err: error }, 'billing_customer_sync.worker_failed')
    }

    if (options?.signal?.aborted) break
    await sleepWithAbort(
      settings.billing.financeProvisioningPollSeconds * 1000,
      options?.signal
    )
  }
}

export function startBillingSyncWorker(options?: { signal?: AbortSignal }): {
  stop: () => void
} {
  const controller = new AbortController()
  const externalSignal = options?.signal

  if (externalSignal) {
    if (externalSignal.aborted) controller.abort()
    else
      externalSignal.addEventListener('abort', () => controller.abort(), {
        once: true,
      })
  }

  void runBillingSyncWorker({ signal: controller.signal })

  return {
    stop: () => controller.abort(),
  }
}
